### PR TITLE
Gradient behavior

### PR DESCRIPTION
**Pull request by Flova**
## Proposed changes
* Adds a gradient behavior
* Adds sidekicks based on local cost map

This is still very WIP

### CODE REVIEW

File: bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py [hunk 1]
```diff
@@ -390,3 +399,111 @@ def get_distance_to_xy(self, x, y):
...
+        fix_points = []
+
+        # Add base points
+        fix_points.extend([
+            # Corner points of the field
+            [[0,            0],             [1,   1]],
+            [[field_length, 0],             [1,   1]],
+            [[0,            field_width],   [1,   1]],
+            [[field_length, field_width],   [1,   1]],
+            # Points in the field that pull the gradient down, so we dont play allways in the middle
```
Comment (johagge) [inline bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py:466]:
```suggestion
            # Points in the field that pull the gradient down, so we don't play always in the middle
```

File: bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py [hunk 2]
```diff
@@ -390,3 +399,111 @@ def get_distance_to_xy(self, x, y):
...
+        plt.show()
+        plt.imshow(interpolated_y.T, origin='lower')
+        plt.show()
+
+    def get_gradient_map_goal(self):
+        position = self.get_ball_position_xy()
+        print(self.gardient_map[0].shape, min(90-1, max(0, int((position[0] + 4.5)) * 10)), int(min(60-1,max(0, (position[1] + 3)) * 10)), self.gardient_map[1][min(9, max(0, int((position[0] + 4.5)))) * 10, int(min(6,max(0, (position[1] + 3)) * 10))], self.gardient_map[0][min(9, max(0, int((position[0] + 4.5)))) * 10, int(min(6,max(0, (position[1] + 3)) * 10))])
+        return math.atan2(
+            self.gardient_map[0][min(90-1, max(0, int((position[0] + 4.5)) * 10)), int(min(60-1,max(0, (position[1] + 3)) * 10))],
+            -self.gardient_map[1][min(90-1, max(0, int((position[0] + 4.5)) * 10)), int(min(60-1,max(0, (position[1] + 3)) * 10))])
```
Comment (johagge) [inline bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py:509]:
```suggestion
        print(self.gradient_map[0].shape, min(90-1, max(0, int((position[0] + 4.5)) * 10)), int(min(60-1,max(0, (position[1] + 3)) * 10)), self.gradient_map[1][min(9, max(0, int((position[0] + 4.5)))) * 10, int(min(6,max(0, (position[1] + 3)) * 10))], self.gradient_map[0][min(9, max(0, int((position[0] + 4.5)))) * 10, int(min(6,max(0, (position[1] + 3)) * 10))])
        return math.atan2(
            self.gradient_map[0][min(90-1, max(0, int((position[0] + 4.5)) * 10)), int(min(60-1,max(0, (position[1] + 3)) * 10))],
            -self.gradient_map[1][min(90-1, max(0, int((position[0] + 4.5)) * 10)), int(min(60-1,max(0, (position[1] + 3)) * 10))])
```

File: bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py [hunk 3]
```diff
@@ -390,3 +399,111 @@ def get_distance_to_xy(self, x, y):
...
         return dist
+
+    ############
+    # Obstacle #
+    ############
+
+    def local_costmap_callback(self, msg):
+        self.local_obstacle_map = msg
+
+    def obstacle_value_at_relative_xy(self, x, y, area=1):
```
Comment (NFiedler) [inline bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py:410]: area is currently not used. I would remove it if it is not necessary as it is the reason the lambda is needed when choosing the kick direction

File: bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py [hunk 4]
```diff
@@ -390,3 +397,161 @@ def get_distance_to_xy(self, x, y):
...
+        y = int((y - 3) * 10)
+        obstacle[x - 2 : x + 2, y - 2: y + 2] = 0
+        self.costmap = self.base_costmap.copy() + gaussian_filter(obstacle, 3)
+        self.calc_gradients()
+
+    def calc_gradients(self):
+        gradient = np.gradient(self.base_costmap)
+        norms = np.linalg.norm(gradient,axis=0)
+        gradient = [np.where(norms==0,0,i/norms) for i in gradient]
+        self.gardient_map = gradient
```
Comment (johagge) [inline bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py:431]: this should be gradient, not gardient

File: bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py [hunk 5]
```diff
@@ -390,3 +397,161 @@ def get_distance_to_xy(self, x, y):
...
+        field_length = rospy.get_param("field_length", 9)
+        x_map = int(min((field_length * 10)-1, max(0, (x + field_length / 2) * 10)))
+        y_map = int(min((field_width * 10)-1,max(0, (y + field_width / 2) * 10)))
+        return -self.gardient_map[0][x_map, y_map], -self.gardient_map[1][x_map, y_map]
+
+    def get_cost_at_field_position(self, x, y):
+        """
+        Gets the costmap value at a given field position
+        :param x: Field coordiante in the x direction
+        :param y: Field coordiante in the y direction
```
Comment (johagge) [inline bitbots_blackboard/src/bitbots_blackboard/capsules/world_model_capsule.py:524]:
```suggestion
        :param x: Field coordinate in the x direction
        :param y: Field coordinate in the y direction
```

**Review by johagge — APPROVED** [review]
I couldn't properly review it, but as you tested it and we need it for the competition, I approved it for now. Someone else should probably review it when we have more time.

**Pull request merged**